mempool/iterators.go

```diff
@@ -0,0 +1,208 @@
+package mempool
+
+import (
+	"fmt"
+
+	"github.com/cometbft/cometbft/internal/clist"
+	"github.com/cometbft/cometbft/types"
+)
+
+// WRRIterator is the base struct for implementing iterators that traverse lanes with
+// the classical Weighted Round Robin (WRR) algorithm.
+type WRRIterator struct {
+	sortedLanes []types.Lane
+	laneIndex   int                            // current lane being iterated; index on sortedLanes
+	counters    map[types.Lane]uint            // counters of consumed entries, for WRR algorithm
+	cursors     map[types.Lane]*clist.CElement // last accessed entries on each lane
+}
+
+func (iter *WRRIterator) nextLane() types.Lane {
+	iter.laneIndex = (iter.laneIndex + 1) % len(iter.sortedLanes)
+	return iter.sortedLanes[iter.laneIndex]
+}
+
+// Non-blocking version of the WRR iterator to be used for reaping and
+// rechecking transactions.
+//
+// Lock must be held on the mempool when iterating: the mempool cannot be
+// modified while iterating.
+type NonBlockingWRRIterator struct {
+	WRRIterator
+}
+
+func NewWRRIterator(mem *CListMempool) *NonBlockingWRRIterator {
+	baseIter := WRRIterator{
+		sortedLanes: mem.sortedLanes,
+		counters:    make(map[types.Lane]uint, len(mem.lanes)),
+		cursors:     make(map[types.Lane]*clist.CElement, len(mem.lanes)),
+	}
+	iter := &NonBlockingWRRIterator{
+		WRRIterator: baseIter,
+	}
+	iter.Reset(mem.lanes)
+	return iter
+}
+
+// Reset must be called before every use of the iterator.
+func (iter *NonBlockingWRRIterator) Reset(lanes map[types.Lane]*clist.CList) {
+	iter.laneIndex = 0
+	for i := range iter.counters {
+		iter.counters[i] = 0
+	}
+	// Set cursors at the beginning of each lane.
+	for lane := range lanes {
+		iter.cursors[lane] = lanes[lane].Front()
+	}
+}
+
+// Next returns the next element according to the WRR algorithm.
+func (iter *NonBlockingWRRIterator) Next() Entry {
+	lane := iter.sortedLanes[iter.laneIndex]
+	numEmptyLanes := 0
+	for {
+		// Skip empty lane or if cursor is at end of lane.
+		if iter.cursors[lane] == nil {
```

Comment:
In `PickLane` (part of the blocking iterator) the condition analogous to this one is more complex, because the fact that the cursor is `nil` doesn't necessarily mean that we are at the end of the lane. Why is it different here?

Reply:
Is it because we don't support changes in the mempool once the iterator has been created?

Reply:
I will try to rethink this again, but the blocking iterator deals with lanes that can be edited while iterating (via recheck etc), here the assumption was the lanes are iterated over when reaping or rechecking , thus locked.

Reply:
But now that you mention it, we should either write down somewhere that this is how it has to be used or change the condition to be the same as in the blocking one. Because this is now a public interface, it can be used in ways we can't foresee now. @hvanz

Reply:
Two reasons for this condition:
- This non-blocking iterator should be used with `updateMtx` locked, so that entries are not added or removed while iterating. There's a comment above the `NonBlockingWRRIterator` struct, maybe we should make it more clear.
- When the iterator is reset (reset is also called when the iterator is created), the cursors are put at the beginning of each lane. So if a lane is empty, its cursor will remain nil. The condition `iter.cursors[lane] == nil` works both for when the lane is empty, or when the cursor reaches the end of the list.

Reply:
I get it. Let me add a suggestion to clarify this